src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializerHelpers.cs

```diff
@@ -0,0 +1,364 @@
+            {
+                string serviceRoot = urlHelper.CreateODataLink(
+                    internalRequest.Context.RouteName,
+                    internalRequest.PathHandler,
+                    new List<ODataPathSegment>());
+
+                try
+                {
+                    odataPath = pathHandler.Parse(serviceRoot, id, internalRequest.RequestContainer).Path;
+                }
+                catch (ODataException)
+                {
+                    return null;
+                }
+            }
+
+            return odataPath;
+        }
+
+        internal static void ApplyODataIdContainer(object resource, ODataResourceWrapper resourceWrapper,
+            ODataDeserializerContext readContext)
+        {
+            Routing.ODataPath path = readContext.Path;
+            if (path == null)
+            {
+                return;
+            }
+
+            IEdmEntityType entityType = path.EdmType.AsElementType() as IEdmEntityType;
+
+            if (entityType != null)
+            {
+                //Setting Odataid , for POCO classes, as a property in the POCO object itself(if user has OdataIDContainer property),
+                //for Delta and EdmEntity object setting as an added property ODataIdcontianer in those classes
+
+                ODataPath ODataPath = new ODataPath(path.Segments);
+
+                // if there is no Id on the resource, try to compute one from path
+                if (resourceWrapper.ResourceBase.Id == null)
+                {
+                    ODataUri odataUri = new ODataUri { Path = ODataPath };
+                    resourceWrapper.ResourceBase.Id = odataUri.BuildUri(ODataUrlKeyDelimiter.Parentheses);
+                }
+
+                if (resourceWrapper.ResourceBase.Id != null)
+                {
+                    string odataId = resourceWrapper.ResourceBase.Id.OriginalString;
+
+                    IODataIdContainer container = new ODataIdContainer();
+                    container.ODataId = odataId;
+
+                    if (resource is EdmEntityObject edmObject)
+                    {
+                        edmObject.ODataIdContainer = container;
+                        edmObject.ODataPath = ODataPath;
+                    }
+                    else if (resource is IDeltaSetItem deltasetItem)
+                    {
+                        deltasetItem.ODataIdContainer = container;
+                        deltasetItem.ODataPath = ODataPath;
+                    }
+                    else
+                    {
+                        PropertyInfo containerPropertyInfo = EdmLibHelpers.GetClrType(entityType, readContext.Model).GetProperties().Where(x => x.PropertyType == typeof(IODataIdContainer)).FirstOrDefault();
+                        if (containerPropertyInfo != null)
+                        {
+                            IODataIdContainer odataIdContainer = containerPropertyInfo.GetValue(resource) as IODataIdContainer;
```

Comment:
I still don't understand... If the property is of type `IODataContainer` - based on the LINQ statement above, why do we need to use `as IODataContainer` here?

Reply:
If my observation/comment above is not valid, please point me to tests I can run to verify each of the scenarios

Reply:
The `GetValue` method returns an object. That's why the casting is done.